On ScheduledDowntime change: remove downtimes created before change

Al2Klimov opened the pull request:
... to cause their re-creation with the ScheduledDowntime change taken into account.

fixes #8309
closes #8630

ref/NC/700763
ref/IP/31583

Al2Klimov commented:
```

```

```
➜  icinga2 git:(feature/scheduleddowntime-change-remove-downtimes-8309) cat 8309.conf
object CheckCommand "dummy" {
	import "dummy-check-command"

	vars.dummy_state = 0
	vars.dummy_text = "Check was successful."
}

object CheckCommand "passive" {
	import "dummy"

	vars.dummy_state = 3
	vars.dummy_text = "No Passive Check Result Received."
}

object Host "localhost" {
	check_command = "passive"
}

apply ScheduledDowntime "lolcat" to Host {
	ranges = {
		"monday - sunday" = "19:00-20:00"
	}
	author = "Hans Meiser"
	comment = "O.O"
	assign where true
}
➜  icinga2 git:(feature/scheduleddowntime-change-remove-downtimes-8309) prefix/sbin/icinga2 daemon -c 8309.conf
[2020-10-05 16:08:57 +0200] information/cli: Icinga application loader (version: v2.12.0-2-g2bc549ea0)
[2020-10-05 16:08:57 +0200] information/cli: Loading configuration file(s).
[2020-10-05 16:08:57 +0200] information/ConfigItem: Committing config item(s).
[2020-10-05 16:08:57 +0200] information/ConfigItem: Instantiated 2 CheckCommands.
[2020-10-05 16:08:57 +0200] information/ConfigItem: Instantiated 1 Host.
[2020-10-05 16:08:57 +0200] information/ConfigItem: Instantiated 1 IcingaApplication.
[2020-10-05 16:08:57 +0200] information/ConfigItem: Instantiated 1 ScheduledDowntime.
[2020-10-05 16:08:57 +0200] information/ScriptGlobal: Dumping variables to file '/Users/aklimov/NET/WS/icinga2/prefix/var/cache/icinga2/icinga2.vars'
[2020-10-05 16:08:57 +0200] information/ConfigItem: Triggering Start signal for config items
[2020-10-05 16:08:57 +0200] information/ConfigItem: Activated all objects.
[2020-10-05 16:09:57 +0200] information/ConfigObjectUtility: Created and activated object 'localhost!29f4daa5-ed24-49c9-90cd-cb3a05364272' of type 'Downtime'.
[2020-10-05 16:09:57 +0200] information/Downtime: Added downtime 'localhost!29f4daa5-ed24-49c9-90cd-cb3a05364272' between '2020-10-05 19:00:00' and '2020-10-05 20:00:00', author: 'Hans Meiser', fixed
^C[2020-10-05 16:10:13 +0200] information/Application: Received request to shut down.
[2020-10-05 16:10:13 +0200] information/Application: Shutting down...
[2020-10-05 16:10:13 +0200] information/ConfigObject: Dumping program state to file '/Users/aklimov/NET/WS/icinga2/prefix/var/lib/icinga2/icinga2.state'
[2020-10-05 16:10:13 +0200] information/IcingaApplication: Icinga has shut down.
➜  icinga2 git:(feature/scheduleddowntime-change-remove-downtimes-8309) perl -pi -e 's/Hans Meiser/Dr jur Hans Meiser/' 8309.conf
➜  icinga2 git:(feature/scheduleddowntime-change-remove-downtimes-8309) prefix/sbin/icinga2 daemon -c 8309.conf
[2020-10-05 16:11:14 +0200] information/cli: Icinga application loader (version: v2.12.0-2-g2bc549ea0)
[2020-10-05 16:11:14 +0200] information/cli: Loading configuration file(s).
[2020-10-05 16:11:14 +0200] information/ConfigItem: Committing config item(s).
[2020-10-05 16:11:14 +0200] information/ConfigItem: Instantiated 2 CheckCommands.
[2020-10-05 16:11:14 +0200] information/ConfigItem: Instantiated 1 Downtime.
[2020-10-05 16:11:14 +0200] information/ConfigItem: Instantiated 1 Host.
[2020-10-05 16:11:14 +0200] information/ConfigItem: Instantiated 1 IcingaApplication.
[2020-10-05 16:11:14 +0200] information/ConfigItem: Instantiated 1 ScheduledDowntime.
[2020-10-05 16:11:14 +0200] information/ScriptGlobal: Dumping variables to file '/Users/aklimov/NET/WS/icinga2/prefix/var/cache/icinga2/icinga2.vars'
[2020-10-05 16:11:14 +0200] information/ConfigObject: Restoring program state from file '/Users/aklimov/NET/WS/icinga2/prefix/var/lib/icinga2/icinga2.state'
[2020-10-05 16:11:14 +0200] information/ConfigObject: Restored 7 objects. Loaded 0 new objects without state.
[2020-10-05 16:11:14 +0200] information/ConfigItem: Triggering Start signal for config items
[2020-10-05 16:11:14 +0200] information/ConfigItem: Activated all objects.
[2020-10-05 16:12:14 +0200] information/ConfigObjectUtility: Deleted object 'localhost!29f4daa5-ed24-49c9-90cd-cb3a05364272' of type 'Downtime'.
[2020-10-05 16:12:14 +0200] information/Downtime: Removed downtime 'localhost!29f4daa5-ed24-49c9-90cd-cb3a05364272' from checkable 'localhost' (Reason: expired at 2020-10-05 20:00:00 +0200).
[2020-10-05 16:13:14 +0200] information/ConfigObjectUtility: Created and activated object 'localhost!d4b7ef11-8035-4b16-ab63-77aa7f128487' of type 'Downtime'.
[2020-10-05 16:13:14 +0200] information/Downtime: Added downtime 'localhost!d4b7ef11-8035-4b16-ab63-77aa7f128487' between '2020-10-05 19:00:00' and '2020-10-05 20:00:00', author: 'Dr jur Hans Meiser', fixed
```

Al2Klimov commented:
Note for the future myself:

* remove all old ones
* but first create new ones
* by considering one as ours only if hash matches
* and by introducing and comparing some timestamps

Al2Klimov commented:
## Blocked by

* [ ] #8768

## Test protocol

```bash
curl -fksSLu root:123456 -X PUT -H 'Accept: application/json' -d '{"attrs":{"author":"me","comment":"mine","ranges":{"wednesday":"20:00-21:00"}}}' 'https://127.0.0.1:5665/v1/objects/scheduleddowntimes/Alexanders-MacBook-Pro.local!1'
```

```
[2021-05-19 17:25:23 +0200] information/ApiListener: New client connection from [::ffff:127.0.0.1]:54460 (no client certificate)
[2021-05-19 17:25:23 +0200] information/ConfigObjectUtility: Created and activated object 'Alexanders-MacBook-Pro.local!1' of type 'ScheduledDowntime'.
[2021-05-19 17:25:23 +0200] information/HttpServerConnection: Request: PUT /v1/objects/scheduleddowntimes/Alexanders-MacBook-Pro.local!1 (from [::ffff:127.0.0.1]:54460), user: root, agent: curl/7.64.1, status: OK).
[2021-05-19 17:25:23 +0200] information/HttpServerConnection: HTTP client disconnected (from [::ffff:127.0.0.1]:54460)
[2021-05-19 17:26:23 +0200] information/ConfigObjectUtility: Created and activated object 'Alexanders-MacBook-Pro.local!0b11ac39-9e19-477c-96c9-49e847767a6a' of type 'Downtime'.
[2021-05-19 17:26:23 +0200] information/Downtime: Added downtime 'Alexanders-MacBook-Pro.local!0b11ac39-9e19-477c-96c9-49e847767a6a' between '2021-05-19 20:00:00' and '2021-05-19 21:00:00', author: 'me', fixed
```

```bash
curl -fksSLu root:123456 -X POST -H 'Accept: application/json' -d '{"attrs":{"author":"me","comment":"mine","ranges":{"wednesday":"18:00-19:00,20:00-21:00"}}}' 'https://127.0.0.1:5665/v1/objects/scheduleddowntimes/Alexanders-MacBook-Pro.local!1'
```

```
[2021-05-19 17:27:13 +0200] information/ApiListener: New client connection from [::ffff:127.0.0.1]:54468 (no client certificate)
[2021-05-19 17:27:13 +0200] information/HttpServerConnection: Request: POST /v1/objects/scheduleddowntimes/Alexanders-MacBook-Pro.local!1 (from [::ffff:127.0.0.1]:54468), user: root, agent: curl/7.64.1, status: OK).
[2021-05-19 17:27:13 +0200] information/HttpServerConnection: HTTP client disconnected (from [::ffff:127.0.0.1]:54468)
[2021-05-19 17:27:21 +0200] information/ConfigObject: Dumping program state to file '/Users/aklimov/NET/WS/icinga2/prefix/var/lib/icinga2/icinga2.state'
[2021-05-19 17:27:23 +0200] information/ConfigObjectUtility: Created and activated object 'Alexanders-MacBook-Pro.local!898a05b6-6538-40b6-9f0e-7e42e6552e7c' of type 'Downtime'.
[2021-05-19 17:27:23 +0200] information/Downtime: Added downtime 'Alexanders-MacBook-Pro.local!898a05b6-6538-40b6-9f0e-7e42e6552e7c' between '2021-05-19 18:00:00' and '2021-05-19 19:00:00', author: 'me', fixed
[2021-05-19 17:27:23 +0200] information/Checkable: Checkable 'Alexanders-MacBook-Pro.local' has 1 notification(s). Checking filters for type 'DowntimeEnd', sends will be logged.
[2021-05-19 17:27:23 +0200] information/Notification: Sending 'DowntimeEnd' notification 'Alexanders-MacBook-Pro.local!mail-icingaadmin' for user 'icingaadmin'
[2021-05-19 17:27:23 +0200] warning/PluginUtility: Error: Non-optional macro 'host.output' used in argument '-o' is missing.
[2021-05-19 17:27:23 +0200] warning/PluginNotificationTask: Notification command for object 'Alexanders-MacBook-Pro.local' (PID: -1, arguments: 'sh' '-c' '') terminated with exit code 3, output: Error: Non-optional macro 'host.output' used in argument '-o' is missing.
[2021-05-19 17:27:23 +0200] information/ConfigObjectUtility: Deleted object 'Alexanders-MacBook-Pro.local!0b11ac39-9e19-477c-96c9-49e847767a6a' of type 'Downtime'.
[2021-05-19 17:27:23 +0200] information/Notification: Completed sending 'DowntimeEnd' notification 'Alexanders-MacBook-Pro.local!mail-icingaadmin' for checkable 'Alexanders-MacBook-Pro.local' and user 'icingaadmin' using command 'mail-host-notification'.
[2021-05-19 17:27:23 +0200] information/Downtime: Removed downtime 'Alexanders-MacBook-Pro.local!0b11ac39-9e19-477c-96c9-49e847767a6a' from checkable 'Alexanders-MacBook-Pro.local' (Reason: expired at 2021-05-19 21:00:00 +0200).
```

* New earlier ranges added – check.
* Old ones removed – check.
* Removal after addition not to cause notification flood – check.

```bash
curl -fksSLu root:123456 -X POST -H 'Accept: application/json' -d '{"attrs":{"author":"me","comment":"mine","ranges":{"wednesday":"18:01-18:59,20:00-21:00"}}}' 'https://127.0.0.1:5665/v1/objects/scheduleddowntimes/Alexanders-MacBook-Pro.local!1'
```

```
[2021-05-19 17:29:23 +0200] information/ConfigObjectUtility: Created and activated object 'Alexanders-MacBook-Pro.local!ae6816d6-13ef-464c-a74b-9dbcdc01794e' of type 'Downtime'.
[2021-05-19 17:29:23 +0200] information/Downtime: Added downtime 'Alexanders-MacBook-Pro.local!ae6816d6-13ef-464c-a74b-9dbcdc01794e' between '2021-05-19 18:01:00' and '2021-05-19 18:59:00', author: 'me', fixed
[2021-05-19 17:29:23 +0200] information/Checkable: Checkable 'Alexanders-MacBook-Pro.local' has 1 notification(s). Checking filters for type 'DowntimeEnd', sends will be logged.
[2021-05-19 17:29:23 +0200] information/Notification: Sending 'DowntimeEnd' notification 'Alexanders-MacBook-Pro.local!mail-icingaadmin' for user 'icingaadmin'
[2021-05-19 17:29:23 +0200] information/ConfigObjectUtility: Deleted object 'Alexanders-MacBook-Pro.local!898a05b6-6538-40b6-9f0e-7e42e6552e7c' of type 'Downtime'.
[2021-05-19 17:29:23 +0200] warning/PluginUtility: Error: Non-optional macro 'host.output' used in argument '-o' is missing.
[2021-05-19 17:29:23 +0200] information/Downtime: Removed downtime 'Alexanders-MacBook-Pro.local!898a05b6-6538-40b6-9f0e-7e42e6552e7c' from checkable 'Alexanders-MacBook-Pro.local' (Reason: expired at 2021-05-19 19:00:00 +0200).
[2021-05-19 17:29:23 +0200] warning/PluginNotificationTask: Notification command for object 'Alexanders-MacBook-Pro.local' (PID: -1, arguments: 'sh' '-c' '') terminated with exit code 3, output: Error: Non-optional macro 'host.output' used in argument '-o' is missing.
[2021-05-19 17:29:23 +0200] information/Notification: Completed sending 'DowntimeEnd' notification 'Alexanders-MacBook-Pro.local!mail-icingaadmin' for checkable 'Alexanders-MacBook-Pro.local' and user 'icingaadmin' using command 'mail-host-notification'.
```

Same for old/new ranges overlapping – check.

Al2Klimov commented:
## Change plan

* re-create everything
* remove only future Downtimes

lippserd reviewed (changes requested):
Please make sure this works without sending notifications and that it also works in a cluster.

Al2Klimov commented:
## Notifications

As we’ve already discussed, we can’t avoid notifications completely. But yes, we should reduce them to the minimum.

While on master...

```
curl -fksSLu root:123456 -X PUT -H 'Accept: application/json' -d '{"attrs":{"author":"me","comment":"mine","ranges":{"tuesday":"19:00-19:30,20:00-21:00"}}}' 'https://127.0.0.1:5665/v1/objects/scheduleddowntimes/Alexanders-MacBook-Pro.local!1'
```

```
[2021-06-29 18:04:02 +0200] information/ApiListener: New client connection from [::ffff:127.0.0.1]:57933 (no client certificate)
[2021-06-29 18:04:02 +0200] information/ConfigObjectUtility: Created and activated object 'Alexanders-MacBook-Pro.local!1' of type 'ScheduledDowntime'.
[2021-06-29 18:04:02 +0200] information/HttpServerConnection: Request: PUT /v1/objects/scheduleddowntimes/Alexanders-MacBook-Pro.local!1 (from [::ffff:127.0.0.1]:57933), user: root, agent: curl/7.64.1, status: OK).
[2021-06-29 18:04:02 +0200] information/HttpServerConnection: HTTP client disconnected (from [::ffff:127.0.0.1]:57933)
```

```
[2021-06-29 18:04:30 +0200] information/ConfigObjectUtility: Created and activated object 'Alexanders-MacBook-Pro.local!4b2cfbab-1ce1-469a-a098-f9e9921510a4' of type 'Downtime'.
[2021-06-29 18:04:30 +0200] information/Downtime: Added downtime 'Alexanders-MacBook-Pro.local!4b2cfbab-1ce1-469a-a098-f9e9921510a4' between '2021-06-29 19:00:00' and '2021-06-29 19:30:00', author: 'me', fixed
```

Then I switch to this PR and wait 5m. All quiet on the western front. The hashes don’t match, but are absent:

```
curl -fksSLu root:123456 -X GET -H 'Accept: application/json' -d '{"pretty":true,"attrs":["scheduled_by","config_owner_hash"]}' 'https://127.0.0.1:5665/v1/objects/downtimes'
{
    "results": [
        {
            "attrs": {
                "config_owner_hash": "",
                "scheduled_by": "Alexanders-MacBook-Pro.local!1"
            },
            "joins": {},
            "meta": {},
            "name": "Alexanders-MacBook-Pro.local!4b2cfbab-1ce1-469a-a098-f9e9921510a4",
            "type": "Downtime"
        },
        {
            "attrs": {
                "config_owner_hash": "",
                "scheduled_by": "Alexanders-MacBook-Pro.local!load!backup-downtime"
            },
            "joins": {},
            "meta": {},
            "name": "Alexanders-MacBook-Pro.local!load!35ad3c56-b936-4617-8e78-592755094e48",
            "type": "Downtime"
        }
    ]
}
```

... until I wipe out all Downtimes:

```
curl -fksSLu root:123456 -X DELETE -H 'Accept: application/json' -d '{"filter":"true"}' 'https://127.0.0.1:5665/v1/objects/downtimes'
```

```
[2021-06-29 18:14:43 +0200] information/ApiListener: New client connection from [::ffff:127.0.0.1]:58043 (no client certificate)
[2021-06-29 18:14:43 +0200] information/Checkable: Checkable 'Alexanders-MacBook-Pro.local' has 1 notification(s). Checking filters for type 'DowntimeEnd', sends will be logged.
[2021-06-29 18:14:43 +0200] information/Notification: Sending 'DowntimeEnd' notification 'Alexanders-MacBook-Pro.local!mail-icingaadmin' for user 'icingaadmin'
[2021-06-29 18:14:43 +0200] information/ConfigObjectUtility: Deleted object 'Alexanders-MacBook-Pro.local!4b2cfbab-1ce1-469a-a098-f9e9921510a4' of type 'Downtime'.
[2021-06-29 18:14:43 +0200] warning/PluginUtility: Error: Non-optional macro 'host.output' used in argument '-o' is missing.
[2021-06-29 18:14:43 +0200] information/Checkable: Checkable 'Alexanders-MacBook-Pro.local!load' has 1 notification(s). Checking filters for type 'DowntimeEnd', sends will be logged.
[2021-06-29 18:14:43 +0200] warning/PluginNotificationTask: Notification command for object 'Alexanders-MacBook-Pro.local' (PID: -1, arguments: 'sh' '-c' '') terminated with exit code 3, output: Error: Non-optional macro 'host.output' used in argument '-o' is missing.
[2021-06-29 18:14:43 +0200] information/Notification: Sending 'DowntimeEnd' notification 'Alexanders-MacBook-Pro.local!load!mail-icingaadmin' for user 'icingaadmin'
[2021-06-29 18:14:43 +0200] information/Notification: Completed sending 'DowntimeEnd' notification 'Alexanders-MacBook-Pro.local!mail-icingaadmin' for checkable 'Alexanders-MacBook-Pro.local' and user 'icingaadmin' using command 'mail-host-notification'.
[2021-06-29 18:14:43 +0200] information/ConfigObjectUtility: Deleted object 'Alexanders-MacBook-Pro.local!load!35ad3c56-b936-4617-8e78-592755094e48' of type 'Downtime'.
[2021-06-29 18:14:43 +0200] warning/PluginUtility: Error: Non-optional macro 'service.output' used in argument '-o' is missing.
[2021-06-29 18:14:43 +0200] information/HttpServerConnection: Request: DELETE /v1/objects/downtimes (from [::ffff:127.0.0.1]:58043), user: root, agent: curl/7.64.1, status: OK).
[2021-06-29 18:14:43 +0200] warning/PluginNotificationTask: Notification command for object 'Alexanders-MacBook-Pro.local!load' (PID: -1, arguments: 'sh' '-c' '') terminated with exit code 3, output: Error: Non-optional macro 'service.output' used in argument '-o' is missing.
[2021-06-29 18:14:43 +0200] information/HttpServerConnection: HTTP client disconnected (from [::ffff:127.0.0.1]:58043)
[2021-06-29 18:14:43 +0200] information/Notification: Completed sending 'DowntimeEnd' notification 'Alexanders-MacBook-Pro.local!load!mail-icingaadmin' for checkable 'Alexanders-MacBook-Pro.local!load' and user 'icingaadmin' using command 'mail-service-notification'.
```

```
[2021-06-29 18:14:47 +0200] information/ConfigObjectUtility: Created and activated object 'Alexanders-MacBook-Pro.local!0ef4d634-2464-4fa8-a88f-1236de7be3fa' of type 'Downtime'.
[2021-06-29 18:14:47 +0200] information/Downtime: Added downtime 'Alexanders-MacBook-Pro.local!0ef4d634-2464-4fa8-a88f-1236de7be3fa' between '2021-06-29 19:00:00' and '2021-06-29 19:30:00', author: 'me', fixed
[2021-06-29 18:14:47 +0200] information/ConfigObjectUtility: Created and activated object 'Alexanders-MacBook-Pro.local!load!fbe58979-3486-4ce1-b966-2f03a480eeef' of type 'Downtime'.
[2021-06-29 18:14:47 +0200] information/Downtime: Added downtime 'Alexanders-MacBook-Pro.local!load!fbe58979-3486-4ce1-b966-2f03a480eeef' between '2021-06-30 02:00:00' and '2021-06-30 03:00:00', author: 'icingaadmin', fixed
```

Now the new ones have hashes:

```
curl -fksSLu root:123456 -X GET -H 'Accept: application/json' -d '{"pretty":true,"attrs":["scheduled_by","config_owner_hash"]}' 'https://127.0.0.1:5665/v1/objects/downtimes'
{
    "results": [
        {
            "attrs": {
                "config_owner_hash": "16c04c1c3fcc679638fc5e5cae535fdb219afd81547ba6d322a30c1b779e06a8",
                "scheduled_by": "Alexanders-MacBook-Pro.local!1"
            },
            "joins": {},
            "meta": {},
            "name": "Alexanders-MacBook-Pro.local!0ef4d634-2464-4fa8-a88f-1236de7be3fa",
            "type": "Downtime"
        },
        {
            "attrs": {
                "config_owner_hash": "ca9502dc8fa5d29c1cb2686808b5d2ccf3ea4a9c6dc3f3c09bfc54614c03c765",
                "scheduled_by": "Alexanders-MacBook-Pro.local!load!backup-downtime"
            },
            "joins": {},
            "meta": {},
            "name": "Alexanders-MacBook-Pro.local!load!fbe58979-3486-4ce1-b966-2f03a480eeef",
            "type": "Downtime"
        }
    ]
}
```

And changes...

```
curl -fksSLu root:123456 -X POST -H 'Accept: application/json' -d '{"attrs":{"author":"me","comment":"mine","ranges":{"tuesday":"19:01-19:30,20:00-21:00"}}}' 'https://127.0.0.1:5665/v1/objects/scheduleddowntimes/Alexanders-MacBook-Pro.local!1'
```

```
[2021-06-29 18:17:21 +0200] information/ApiListener: New client connection from [::ffff:127.0.0.1]:58065 (no client certificate)
[2021-06-29 18:17:21 +0200] information/HttpServerConnection: Request: POST /v1/objects/scheduleddowntimes/Alexanders-MacBook-Pro.local!1 (from [::ffff:127.0.0.1]:58065), user: root, agent: curl/7.64.1, status: OK).
[2021-06-29 18:17:21 +0200] information/HttpServerConnection: HTTP client disconnected (from [::ffff:127.0.0.1]:58065)
```

... are forwarded:

```
[2021-06-29 18:17:47 +0200] information/ConfigObjectUtility: Created and activated object 'Alexanders-MacBook-Pro.local!ccdf3ef5-bcf1-4fc4-9e92-d2444d8134a6' of type 'Downtime'.
[2021-06-29 18:17:47 +0200] information/Downtime: Added downtime 'Alexanders-MacBook-Pro.local!ccdf3ef5-bcf1-4fc4-9e92-d2444d8134a6' between '2021-06-29 19:01:00' and '2021-06-29 19:30:00', author: 'me', fixed
[2021-06-29 18:17:47 +0200] information/Checkable: Checkable 'Alexanders-MacBook-Pro.local' has 1 notification(s). Checking filters for type 'DowntimeEnd', sends will be logged.
[2021-06-29 18:17:47 +0200] information/Notification: Sending 'DowntimeEnd' notification 'Alexanders-MacBook-Pro.local!mail-icingaadmin' for user 'icingaadmin'
[2021-06-29 18:17:47 +0200] information/ConfigObjectUtility: Deleted object 'Alexanders-MacBook-Pro.local!0ef4d634-2464-4fa8-a88f-1236de7be3fa' of type 'Downtime'.
[2021-06-29 18:17:47 +0200] warning/PluginUtility: Error: Non-optional macro 'host.output' used in argument '-o' is missing.
[2021-06-29 18:17:47 +0200] information/Downtime: Removed downtime 'Alexanders-MacBook-Pro.local!0ef4d634-2464-4fa8-a88f-1236de7be3fa' from checkable 'Alexanders-MacBook-Pro.local' (Reason: expired at 2021-06-29 19:30:00 +0200).
[2021-06-29 18:17:47 +0200] warning/PluginNotificationTask: Notification command for object 'Alexanders-MacBook-Pro.local' (PID: -1, arguments: 'sh' '-c' '') terminated with exit code 3, output: Error: Non-optional macro 'host.output' used in argument '-o' is missing.
[2021-06-29 18:17:47 +0200] information/Notification: Completed sending 'DowntimeEnd' notification 'Alexanders-MacBook-Pro.local!mail-icingaadmin' for checkable 'Alexanders-MacBook-Pro.local' and user 'icingaadmin' using command 'mail-host-notification'.
```

Al2Klimov commented:
## Cluster

Sure, older versions on satellites don’t know the new attribute and _would_ fail to validate the config. Fortunately:

* Downtimes stuck on a cluster level
* Downtimes are created w/ `ignore_on_error`

julianbrost commented on lib/icinga/scheduleddowntime.cpp:
```diff
@@ -292,6 +310,22 @@ void ScheduledDowntime::CreateNextDowntime()
+{
+	auto name (GetName());
+	auto downtimeOptionsHash (HashDowntimeOptions());
+	auto threshold (Utility::GetTime() + 5 * 60); // just to be sure
```
Be sure about what?

julianbrost commented:
Regarding cluster sync of downtimes: If a downtime is scheduled from a scheduled downtime object, the downtime object inherits the zone of the scheduled downtime (otherwise it inherits the one of the checkable), so downtimes from scheduled ones should never leave the zone that scheduled them:

https://github.com/Icinga/icinga2/blob/e180241e4a8b92da9554272b7d49c07d7c2f808e/lib/icinga/downtime.cpp#L259-L278

julianbrost commented on lib/icinga/scheduleddowntime.cpp:
```diff
@@ -292,6 +310,24 @@ void ScheduledDowntime::CreateNextDowntime()
+		if (downtime->GetScheduledBy() == name && downtime->GetStartTime() > threshold) {
+			auto configOwnerHash (downtime->GetConfigOwnerHash());
+
+			if (!configOwnerHash.IsEmpty() && configOwnerHash != downtimeOptionsHash)
```
What was the intention of not re-creating downtimes that don't have the config owner hash set?

Al2Klimov replied:
Imagine you have ScheduledDowntimes and DowntimeStart notifications everywhere. And then you upgrade...

julianbrost replied:
But since you only delete downtimes that start in more than 5 minutes, no start notifications have been sent yet. So I think the following should work fine:
1. Remove the `!configOwnerHash.IsEmpty()` here -> delete every downtime with no or a wrong hash starting more that 5 minutes in the future
2. Leave the check in `CreateNextDowntime()` as it is right now -> don't reschedule if there's still a downtime without a hash set
3. In `TimerProc()`, call `RemoveObsoleteDowntimes()` before `CreateNextDowntime()` to minimize the time between deleting the old and adding the new downtime

Al2Klimov replied:
I'm afraid that doesn’t matter for the end notification.

```
[2021-07-02 17:05:51 +0200] information/ConfigObjectUtility: Created and activated object 'Alexanders-MacBook-Pro.local!48db09bb-1ea6-40d1-8bac-d56c57ddfd15' of type 'Downtime'.
[2021-07-02 17:05:51 +0200] information/Downtime: Added downtime 'Alexanders-MacBook-Pro.local!48db09bb-1ea6-40d1-8bac-d56c57ddfd15' between '2021-07-02 17:07:00' and '2021-07-02 19:30:00', author: 'me', fixed
[2021-07-02 17:05:51 +0200] information/Checkable: Checkable 'Alexanders-MacBook-Pro.local' has 1 notification(s). Checking filters for type 'DowntimeEnd', sends will be logged.
[2021-07-02 17:05:51 +0200] information/Notification: Sending 'DowntimeEnd' notification 'Alexanders-MacBook-Pro.local!mail-icingaadmin' for user 'icingaadmin'
[2021-07-02 17:05:51 +0200] information/ConfigObjectUtility: Deleted object 'Alexanders-MacBook-Pro.local!11f091fb-d434-48d7-bb90-4aa76614dfcd' of type 'Downtime'.
[2021-07-02 17:05:51 +0200] information/Notification: Completed sending 'DowntimeEnd' notification 'Alexanders-MacBook-Pro.local!mail-icingaadmin' for checkable 'Alexanders-MacBook-Pro.local' and user 'icingaadmin' using command 'mail-host-notification'.
[2021-07-02 17:05:51 +0200] information/Downtime: Removed downtime 'Alexanders-MacBook-Pro.local!11f091fb-d434-48d7-bb90-4aa76614dfcd' from checkable 'Alexanders-MacBook-Pro.local' (Reason: expired at 2021-07-02 19:30:00 +0200).
```

julianbrost replied:
Wait what, end notifications are also sent for downtimes that weren't even started?

julianbrost replied:
Should be fixed by #8874

julianbrost commented on lib/icinga/scheduleddowntime.cpp:
```diff
@@ -236,11 +249,16 @@ void ScheduledDowntime::CreateNextDowntime()
 		if (downtime->GetScheduledBy() != GetName())
 			continue;
 
+		auto configOwnerHash (downtime->GetConfigOwnerHash());
+		if (!configOwnerHash.IsEmpty() && configOwnerHash != downtimeOptionsHash)
+			continue;
```
So downtimes that don't have a config owner hash set will still be considered for `minEnd`. Therefore, I think if you add a range that's earlier than the next one already scheduled, it won't reschedule. Thus, probably won't fix #8630 in the current state.

Al2Klimov replied:
The fact you mentioned in the review above already makes this construct temporarily somewhat broken. Users will have to let it run for a while (until the old DTs vanish).

Al2Klimov commented:
## Blocked by

* #8874

julianbrost commented:
Is it only blocked by that PR or do we still want to change this PR to also reschedule existing downtimes that don't have the hash set once that's fixed?

Al2Klimov commented:
I consider the latter not necessary.

julianbrost commented:
Is there any technical reason not to do it (i.e. any wrong notifications or similar)?

Al2Klimov commented:
Less fluctuations.

julianbrost commented:
What kind of fluctuations? If you can think of anything that might be a problem there, we should ensure that won't be a problem when the config is actually changed.

Al2Klimov commented:
Imagine you upgrade. And 90% Downtimes are re-created.

julianbrost commented:
But will this cause any actual problems? Imagine you update your config and 90% of all downtimes are recreated.

Al2Klimov commented:
As far as I remember we already agreed on _users will have to let the old DTs expire_ .

julianbrost commented:
If I remember correctly, that decision was based on the claim that doing so would generate notifications (we can still stick to that decision nonetheless). Can you please give a simple yes/no answer to the following question: If I update the config of all existing ScheduleDowntimes (which should have the same effect as if we'd reschedule all downtimes once with the upgrade), can you imagine any problems caused due to this?

Al2Klimov commented:
No.

julianbrost commented:
From reading the code (still have to spin it up for a test), the current state of the PR should give the following behavior:
* All downtime objects created before this code change are kept as until their time passes, changing the configuration will have no effect on these and not reschedule.
* Otherwise (after this code created downtime objects for a ScheduledDowntime), when the config is changed, one of two things may happen:
  * If the downtime created using the old config already started or will start within the next 5 minutes, the old downtime object is kept but one using the new config will also be created (those two downtimes may overlap)
  * If the old downtime starts in more than 5 minutes, a new one using the new config is created and the old one is discarded.

With #8874, none of these operations should generate any notifications.

lippserd commented:
We also discussed not to recreate existing downtimes, as it could hit quite a lot of downtimes at once, slowing down Icinga unnecessarily.

julianbrost reviewed (approved):
Works as described in https://github.com/Icinga/icinga2/pull/8310#issuecomment-874844767
